File: frontend/src/services/tank-transaction-service.ts
import api from './api';
import type {
  TankTransactionResponse,
  CreateTankTransactionRequest,
} from '@/types';

export class TankTransactionService {
  private static readonly BASE_PATH = '/api/v1/tanks';

  // Get transactions for a tank
  static async getTransactions(
    tankId: string
  ): Promise<TankTransactionResponse[]> {
    const response = await api.get<TankTransactionResponse[]>(
      `${this.BASE_PATH}/${tankId}/transactions`
    );
    return response.data;
  }

  // Get transactions for a tank with date range
  static async getTransactionsWithDateRange(
    tankId: string,
    fromDate: string,
    toDate: string
  ): Promise<TankTransactionResponse[]> {
    const queryParams = new URLSearchParams();
    queryParams.append('fromDate', fromDate);
    queryParams.append('toDate', toDate);

    const response = await api.get<TankTransactionResponse[]>(
      `${this.BASE_PATH}/${tankId}/transactions?${queryParams}`
    );
    return response.data;
  }

  // Create addition transaction (fuel added to tank)
  static async createAddition(
    tankId: string,
    transaction: CreateTankTransactionRequest
  ): Promise<TankTransactionResponse> {
    const response = await api.post<TankTransactionResponse>(
      `${this.BASE_PATH}/${tankId}/addition`,
      transaction
    );
    return response.data;
  }

  // Create removal transaction (fuel removed from tank)
  static async createRemoval(
    tankId: string,
    transaction: CreateTankTransactionRequest
  ): Promise<TankTransactionResponse> {
    const response = await api.post<TankTransactionResponse>(
      `${this.BASE_PATH}/${tankId}/removal`,
      transaction
    );
    return response.data;
  }

  // Get opening level for a specific date
  static async getOpeningLevel(tankId: string, date: string): Promise<number> {
    const queryParams = new URLSearchParams();
    queryParams.append('date', date);

    const response = await api.get<number>(
      `${this.BASE_PATH}/${tankId}/opening-level?${queryParams}`
    );
    return response.data;
  }
}

File: backend/src/main/java/com/reallink/pump/services/TankTransactionService.java
package com.reallink.pump.services;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.validation.annotation.Validated;

import com.reallink.pump.dto.request.CreateTankTransactionRequest;
import com.reallink.pump.dto.response.TankTransactionResponse;
import com.reallink.pump.entities.DailyTankLevel;
import com.reallink.pump.entities.Tank;
import com.reallink.pump.entities.TankTransaction;
import com.reallink.pump.entities.TankTransaction.TransactionType;
import com.reallink.pump.exception.PumpBusinessException;
import com.reallink.pump.mapper.TankTransactionMapper;
import com.reallink.pump.repositories.DailyTankLevelRepository;
import com.reallink.pump.repositories.TankRepository;
import com.reallink.pump.repositories.TankTransactionRepository;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import lombok.RequiredArgsConstructor;

@Service
@Validated
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class TankTransactionService {

    private final TankTransactionRepository transactionRepository;
    private final TankRepository tankRepository;
    private final TankTransactionMapper mapper;
    private final DailyTankLevelRepository dailyTankLevelRepository;

    public List<TankTransactionResponse> getTransactionsByTankId(@NotNull UUID tankId) {
        List<TankTransaction> transactions = transactionRepository.findByTankIdOrderByTransactionDateDesc(tankId);
        return mapper.toResponseList(transactions);
    }

    public List<TankTransactionResponse> getTransactionsByTankIdAndDateRange(
            UUID tankId, LocalDate fromDate, LocalDate toDate) {
        List<TankTransaction> transactions = transactionRepository
                .findByTankIdAndTransactionDateBetweenOrderByTransactionDateAsc(
                        tankId, fromDate.atStartOfDay(), toDate.atTime(23, 59, 59));
        return transactions.stream()
                .map(mapper::toResponse)
                .collect(Collectors.toList());
    }

    @Transactional
    public TankTransactionResponse createAdditionTransaction(@NotNull UUID tankId, @Valid CreateTankTransactionRequest request) {
        Tank tank = tankRepository.findById(tankId).orElse(null);
        if (tank == null) {
            throw new PumpBusinessException("TANK_NOT_FOUND", "Tank with ID " + tankId + " not found");
        }
        TankTransaction transaction = mapper.toEntity(request);
        transaction.setTank(tank);
        transaction.setTransactionType(TransactionType.ADDITION);
        transaction.setEntryBy(SecurityContextHolder.getContext().getAuthentication().getName());
        if (transaction.getTransactionDate() == null) {
            transaction.setTransactionDate(LocalDateTime.now());
        }
        TankTransaction saved = transactionRepository.save(transaction);
        updateDailyTankLevel(saved);
        return mapper.toResponse(saved);
    }

    @Transactional
    public TankTransactionResponse createRemovalTransaction(@NotNull UUID tankId, @Valid CreateTankTransactionRequest request) {
        Tank tank = tankRepository.findById(tankId).orElse(null);
        if (tank == null) {
            throw new PumpBusinessException("TANK_NOT_FOUND", "Tank with ID " + tankId + " not found");
        }
        TankTransaction transaction = mapper.toEntity(request);
        transaction.setTank(tank);
        transaction.setTransactionType(TransactionType.REMOVAL);
        transaction.setEntryBy(SecurityContextHolder.getContext().getAuthentication().getName());
        if (transaction.getTransactionDate() == null) {
            transaction.setTransactionDate(LocalDateTime.now());
        }
        TankTransaction saved = transactionRepository.save(transaction);
        updateDailyTankLevel(saved);
        return mapper.toResponse(saved);
    }

    private void updateDailyTankLevel(TankTransaction transaction) {
        LocalDate date = transaction.getTransactionDate().toLocalDate();
        BigDecimal dailyNet = transactionRepository.getDailyNetByTankIdAndDate(transaction.getTank().getId(), date);
        DailyTankLevel dailyLevel = dailyTankLevelRepository.findByTank_IdAndDate(transaction.getTank().getId(), date)
                .orElse(new DailyTankLevel());
        dailyLevel.setTank(transaction.getTank());
        dailyLevel.setDate(date);
        dailyLevel.setDailyNet(dailyNet);
        dailyTankLevelRepository.save(dailyLevel);
    }

    public BigDecimal getOpeningLevel(UUID tankId, LocalDate date) {
        Tank tank = tankRepository.findById(tankId).orElse(null);
        if (tank == null) {
            throw new PumpBusinessException("TANK_NOT_FOUND", "Tank with ID " + tankId + " not found");
        }

        // Get cumulative additions up to the specified date
        BigDecimal cumulativeAdditions = dailyTankLevelRepository.getCumulativeNetUpToDate(tankId, date);

        // Return opening level + cumulative additions
        return tank.getOpeningLevel().add(cumulativeAdditions);
    }
}
